### PR TITLE
Make some awaits safer and add a guard on cloning the repo

### PR DESCRIPTION
**Pull request by blowery**
Some awaits were missing try blocks so any errors they generated ended up unhandled. Add some trys and catches with logging.
Also, it was possible to kick off the clone more than one against the same target directory, or try to open the target directory while the clone was ongoing. Add a promise gate around that process to make it safe.

### CODE REVIEW

File: src/api.ts [hunk 1]
```diff
@@ -155,6 +167,53 @@ export function getPortForContainer(hash: CommitHash): number|boolean {
 		: false;
 }
 
+var repoClonePromises: Map<string, Promise<git.Repository> > = new Map();
```
Comment (dmsnell) [inline src/api.ts]: do you not want these in `APIState`?

Comment (blowery) [reply src/api.ts]: what's the benefit?

Comment (dmsnell) [reply src/api.ts]: I moved a bunch of things into one state atom to make it easier for runtime introspection. we don't have the display yet, but shoving all the state together is easier for my poor brain to comprehend

**Comment by blowery** [comment]
Not sure if this would solve it, but i've been seeing a lot of "branch missing" errors while building a branch on dserve. A refresh usually clears it up. Feels like some kind of race fetching the branch info.

**Comment by sirreal** [comment]
Is there work here worth refreshing, reviewing, and shipping? I'll do my best to help if so.

**Comment by blowery** [comment]
@sirreal nah, I think we're ok at the moment. It's been pretty stable.

**Pull request closed without merging**